executor/aggfuncs/func_max_min.go

```diff
@@ -14,121 +14,133 @@
+	if len(d.Items) <= 0 {
+		return nil, errors.New("Pop front when Deque is empty")
 	}
+	res := d.Items[0]
+	d.Items = d.Items[1:]
+	return res, nil
 }
-func (h *maxMinHeap) Remove(val interface{}) {
-	if h.varSet[val] > 0 {
-		h.varSet[val]--
-	} else {
-		panic("remove a not exist value")
+
+// PopBack pops an Item from the end of Deque
+func (d *Deque) PopBack() (interface{}, error) {
+	i := len(d.Items) - 1
+	if i < 0 {
+		return nil, errors.New("Pop back when Deque is empty")
 	}
+	res := d.Items[i]
+	d.Items = d.Items[:i]
+	return res, nil
 }
-func (h *maxMinHeap) Top() (val interface{}, isEmpty bool) {
-retry:
-	if h.Len() == 0 {
-		return nil, true
-	}
-	top := h.data[0]
-	if h.varSet[top] == 0 {
-		_ = heap.Pop(h)
-		goto retry
+
+// Back returns the element at the end of Deque
+func (d *Deque) Back() (Pair, error) {
+	i := len(d.Items) - 1
+	if i < 0 {
+		return Pair{}, errors.New("Cannot get back when Deque is empty")
 	}
-	return top, false
+	return d.Items[i], nil
 }
 
-func (h *maxMinHeap) AppendMyDecimal(val types.MyDecimal) error {
-	key, err := val.ToHashKey()
-	if err != nil {
-		return err
-	}
-	h.varSet[string(key)]++
-	if h.varSet[string(key)] == 1 {
-		heap.Push(h, val)
+// Front returns the element at the front of Deque
+func (d *Deque) Front() (Pair, error) {
+	if len(d.Items) <= 0 {
+		return Pair{}, errors.New("Cannot get front when Deque is empty")
 	}
-	return nil
+	return d.Items[0], nil
 }
-func (h *maxMinHeap) RemoveMyDecimal(val types.MyDecimal) error {
-	key, err := val.ToHashKey()
-	if err != nil {
-		return err
-	}
-	if h.varSet[string(key)] > 0 {
-		h.varSet[string(key)]--
-	} else {
-		panic("remove a not exist value")
+
+// IsEmpty returns if Deque is empty
+func (d *Deque) IsEmpty() bool {
+	return len(d.Items) == 0
+}
+
+// Pair pairs items and their indices in Deque
+type Pair struct {
+	Item interface{}
+	Idx  uint64
+}
+
+// Reset Deque
+func (d *Deque) Reset() {
+	d.Items = d.Items[:0]
+}
+
+// Dequeue pops out element from the front, if element's index is out of boundary, i.e. the leftmost element index
+func (d *Deque) Dequeue(boundary uint64) error {
+	for !d.IsEmpty() {
+		frontEle, err := d.Front()
+		if err != nil {
+			return err
+		}
+		if frontEle.Idx <= boundary {
+			_, err := d.PopFront()
+			if err != nil {
+				return err
+			}
+			continue
+		}
+		return nil
 	}
 	return nil
 }
-func (h *maxMinHeap) TopDecimal() (val types.MyDecimal, isEmpty bool) {
-retry:
-	if h.Len() == 0 {
-		return types.MyDecimal{}, true
-	}
-	top := h.data[0].(types.MyDecimal)
-	key, err := top.ToHashKey()
-	if err != nil {
-		panic(err)
-	}
-	if h.varSet[string(key)] == 0 {
-		_ = heap.Pop(h)
-		goto retry
+
+// Enqueue put Item at the back of queue, while popping any element that is lesser element in queue
+func (d *Deque) Enqueue(idx uint64, item interface{}) error {
+	for !d.IsEmpty() {
+		pair, err := d.Back()
+		if err != nil {
+			return err
+		}
+
+		bigger := d.cmpFunc(item, pair.Item) >= 0
```

Comment:
```
cmp := d.cmpFunc(item, pair.Item)
if cmp >= 0 && d.IsMax || cmp <=0 && !d.IsMax {
...
}
```
I think the logic is better, since it can eliminate the values that equals when d.IsMax is false.

Reply:
Agreed, looks more concise now.